### ISSUE TITLE
Slightly incorrect error on mutable borrow of immutable captured variable

### ISSUE THREAD

**Issue by bkoropoff** [rust-lang/rust#18330]
Code:

``` rust
fn set(x: &mut usize) { *x = 1; }

fn main() {
    let x = 0;
    || set(&mut x);
}
```

Output:

```
<anon>:7:5: 7:19 error: closure cannot assign to immutable local variable `x`
<anon>:7     || set(&mut x);
             ^~~~~~~~~~~~~~
```

This should actually state that it cannot borrow (rather than assign) the variable.  This has been around since 0.10 at least.

**Comment by steveklabnik** [comment]
Triage: no change today.

**Comment by cyplo** [comment]
Hello ! 
The message seems to be different now, is it what you would expect, @bkoropoff ?

```
error: closure cannot assign to immutable local variable `x`
 --> test.rs:5:5
  |
4 |     let x = 0;
  |         - consider changing this to `mut x`
5 |     || set(&mut x);
  |     ^^ cannot borrow mutably
```

`rustc 1.18.0-nightly (28a742997 2017-04-13)`

**Comment by JakubAdamWieczorek** [comment]
`rustc 1.22.0-nightly (981ce7d8d 2017-09-03)`

Input:
```rust
fn f(x: Option<()>) {
	(|| x.take())();
}
fn main() {}
```

Output:
```
error[E0595]: closure cannot assign to immutable argument `x`
 --> test.rs:2:3
  |
1 | fn f(x: Option<()>) {
  |      - consider changing this to `mut x`
2 |     (|| x.take())();
  |   ^^ cannot borrow mutably

error: aborting due to previous error
```

Still not quite. `f()` doesn't really reassign the argument, it just calls a function that borrows mutably. It's really just a matter of a slight wording change to the error message to make it more general. Much like in the PR https://github.com/rust-lang/rust/pull/29670, which ended up not being merged.

**Comment by estebank** [comment]
From #45622:

The output should point out at the use of `x` in some way, as well as changing the wording slightly (it is not an assignment):

```
error[E0595]: closure cannot modify immutable argument `x`
 --> src/main.rs:4:6
  |
1 | fn foo(x: &mut i32) {
  |        - consider changing this to `mut x`
...
4 |     (|| {
  |      ^^ cannot borrow mutably
5 |        process(&mut x);
  |                     - `x` used here
```
